core/src/main/java/org/elasticsearch/common/settings/SecureString.java

```diff
@@ -106,8 +106,29 @@ public synchronized String toString() {
      */
     @Override
     public synchronized void close() {
-        Arrays.fill(chars, '\0');
-        chars = null;
+        if (chars != null) {
+            Arrays.fill(chars, '\0');
+            chars = null;
+        }
+    }
+
+    /**
+     * Returns a closeable object that provides access to an array of chars that is a copy of this SecureString's backing chars. Upon
+     * closing this object the char array will be cleared without affecting this SecureString. This is useful for APIs which accept a
+     * char[]. For example:
+     *
+     * <pre>
+     *     try (CloseableChars closeableChars = secureString.getAsCloseableChars()) {
```

Comment:
> I'm confused on why you wouldn't just close the secureString?

I thought we were going to clear the value out of the in memory keystore at some point once we change the code to only read these during initialization (from Bootstrap.java):

```java
           /* TODO: close this once s3 repository doesn't try to read during repository construction
            try {
                // any secure settings must be read during node construction
                IOUtils.close(keystore);
            } catch (IOException e) {
                throw new BootstrapException(e);
            }*/
```

The reasoning behind the additional class is that it provides a way to pass a copy of the chars to another consumer and prevents this consumer from manipulating the backing char[] of the secure string, while also providing an easy way to clear the array once it is finished being used. The secure string may need to be used in multiple places so I do not think it is safe to pass the only copy of the char[] to other code.

Maybe a simpler option would be to add a method to get the underlying chars and another method that returns a new copy of the secure string? Then the code would be something like:

```java
try (SecureString copy = secureString.newCopy()) {
    // execute code that uses the char[] one time and no longer needs it such as the keystore API
    KeyStore store = KeyStore.getInstance("jks");
    store.load(inputStream, copy.getChars());
}
```